### PR TITLE
[RFP]socks5 support for JsonRpsee

### PR DESCRIPTION
**Pull request by flipchan**
This is an RFP for adding support for socks5 proxy connections in the Rust library [JsonRpsee](https://github.com/paritytech/jsonrpsee).

The RFP is created in order to motivate people to work on pr https://github.com/paritytech/jsonrpsee/issues/870.

By enabling socks5 support several ecosystem client libraries can connect to .onion(tor) rpc endpoints. Giving end users more privacy when connecting to nodes.

### CODE REVIEW

**Review by Noc2 — APPROVED** [review]
Thanks a lot for creating the RFP. I would be happy to merge it.

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Socks5 proxy support for JsonRpsee 
+
+* **Status:** Open 
+* **Proposer:** flipchan   
+
+## Project Description :page_facing_up: 
+
+[JsonRpsee](https://github.com/paritytech/jsonrpsee) is an opensource Rust library that is used by several para and relay chain client libraries to connect to chains over ws or rpc.
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:8]:
```suggestion
[JsonRpsee](https://github.com/paritytech/jsonrpsee) is an opensource Rust library used by several para and relay chain client libraries to connect to chains over ws or rpc. 
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+* **Status:** Open 
+* **Proposer:** flipchan   
+
+## Project Description :page_facing_up: 
+
+[JsonRpsee](https://github.com/paritytech/jsonrpsee) is an opensource Rust library that is used by several para and relay chain client libraries to connect to chains over ws or rpc. 
+
+## Background story:  
+
+Tor is a anonymization project that has been active for several years. By utilizing a traffic obfusication protocol to route data, it enables users to anonymize their internet traffic. Several blockchain projects leverage this technology to allow users to connect to nodes over tor, inputing the .onion address of the rpc provider.
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:12]:
```suggestion
Tor is a anonymization project that has been active for several years. By utilizing a traffic obfusication protocol to route data, it enables users to anonymize their internet traffic. Several blockchain projects leverage this technology to allow users to connect to nodes over tor, inputting the .onion address of the rpc provider.  
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+## Project Description :page_facing_up: 
+
+[JsonRpsee](https://github.com/paritytech/jsonrpsee) is an opensource Rust library that is used by several para and relay chain client libraries to connect to chains over ws or rpc. 
+
+## Background story:  
+
+Tor is a anonymization project that has been active for several years. By utilizing a traffic obfusication protocol to route data, it enables users to anonymize their internet traffic. Several blockchain projects leverage this technology to allow users to connect to nodes over tor, inputing the .onion address of the rpc provider.  
+
+By providing a .onion address for a rpc/ws node, we can allow users to have more privacy by masking the origination of the request(senders ip address).
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:14]:
```suggestion
By providing a .onion address for a rpc/ws node, we can allow users to have more privacy by masking the origination of the request (the sender's ip address).   
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+[JsonRpsee](https://github.com/paritytech/jsonrpsee) is an opensource Rust library that is used by several para and relay chain client libraries to connect to chains over ws or rpc. 
+
+## Background story:  
+
+Tor is a anonymization project that has been active for several years. By utilizing a traffic obfusication protocol to route data, it enables users to anonymize their internet traffic. Several blockchain projects leverage this technology to allow users to connect to nodes over tor, inputing the .onion address of the rpc provider.  
+
+By providing a .onion address for a rpc/ws node, we can allow users to have more privacy by masking the origination of the request(senders ip address).   
+
+
+Gavin Wood mentioning onion-routing as an important future on Decoded 2023:
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:17]:
```suggestion
Gavin Wood mentioned onion-routing as an important future feature on Decoded 2023:
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+In order to connect to a .onion site, a user must pass it's connection through a tor socks5 proxy in order to resolve the .onion domain and connect.  
+
+
+Several ecosystem projects want to add support for connecting to .onion, but are blocked due to JsonRpsee not having support for sock5 proxy.    
+
+
+Third party pr's that are waiting for JsonRpsee to support socks5:  
+[Fedimint #391](https://github.com/fedimint/fedimint/issues/391#issuecomment-1643878929)  
+[Uptest #15](https://github.com/uptest-sc/uptest/issues/15)
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:32]:
```suggestion
Third party pr's that are waiting for JsonRpsee to support socks5:
- [Fedimint #391](https://github.com/fedimint/fedimint/issues/391#issuecomment-1643878929)  
- [Uptest #15](https://github.com/uptest-sc/uptest/issues/15)   
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+Several ecosystem projects want to add support for connecting to .onion, but are blocked due to JsonRpsee not having support for sock5 proxy.    
+
+
+Third party pr's that are waiting for JsonRpsee to support socks5:  
+[Fedimint #391](https://github.com/fedimint/fedimint/issues/391#issuecomment-1643878929)  
+[Uptest #15](https://github.com/uptest-sc/uptest/issues/15)   
+
+On 4th of September of 2022 a [pr was created](https://github.com/paritytech/jsonrpsee/issues/870) to start the process of adding socks5 support for JsonRpsee.  
+Noone has had time to fix this issue and implement this feature, therefor this RFP.
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:35]:
```suggestion
No one has had time to fix this issue and implement this feature, therefore this RFP. 
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+* **Full-time equivalent (FTE):**  4 weeks   
+* **Total Costs:** 9000 USD(may be changed by the future team) 
+
+
+### Milestone 1
+
+Please add additional milestones in the same way: 
+* **Estimated Duration:** Duration of milestone 1 
+* **FTE:**  4 weeks
+* **Costs:** 9000 USD(may be changed by the future team)
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:61]:
```suggestion
* **Costs:** 9000 USD (may be changed by the future team)
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+-  [Substrate](https://github.com/paritytech/substrate)  
+-  [Subwasm](https://github.com/chevdor/subwasm)  
+-  [Uptest](https://github.com/uptest-sc/uptest)   
+-  [Subway](https://github.com/AcalaNetwork/subway)  
+
+## Deliverables :nut_and_bolt:
+
+* **Total Estimated Duration:** 4 weeks  
+* **Full-time equivalent (FTE):**  4 weeks   
+* **Total Costs:** 9000 USD(may be changed by the future team)
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:53]:
```suggestion
* **Total Costs:** 9000 USD (may be changed by the future team) 
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
+# Socks5 proxy support for JsonRpsee 
+
+* **Status:** Open 
+* **Proposer:** flipchan   
+
+## Project Description :page_facing_up: 
+
+[JsonRpsee](https://github.com/paritytech/jsonrpsee) is an opensource Rust library that is used by several para and relay chain client libraries to connect to chains over ws or rpc. 
+
+## Background story:
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:10]:
```suggestion
## Background story
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+Third party pr's that are waiting for JsonRpsee to support socks5:  
+[Fedimint #391](https://github.com/fedimint/fedimint/issues/391#issuecomment-1643878929)  
+[Uptest #15](https://github.com/uptest-sc/uptest/issues/15)   
+
+On 4th of September of 2022 a [pr was created](https://github.com/paritytech/jsonrpsee/issues/870) to start the process of adding socks5 support for JsonRpsee.  
+Noone has had time to fix this issue and implement this feature, therefor this RFP. 
+
+
+## Motivation:
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:38]:
```suggestion
## Motivation
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+On 4th of September of 2022 a [pr was created](https://github.com/paritytech/jsonrpsee/issues/870) to start the process of adding socks5 support for JsonRpsee.  
+Noone has had time to fix this issue and implement this feature, therefor this RFP. 
+
+
+## Motivation:  
+-  Enable client libraries to connect to .onion rpc nodes.   
+
+
+### Ecosystem projects that rely on JsonRpsee:
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:42]:
```suggestion
### Ecosystem projects that rely on JsonRpsee
```

File: docs/RFPs/jsonrpsee-proxy-support.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+| ------------- | ------------- | ------------- |
+| 0a.  | License |	Apache 2.0 / MIT / Unlicense| 
+| 0b.  | Documentation | Public documentation with implementation guides and sample code | 
+| 0c.  | Testing Guide |Rust Unit tests| 
+| 0d.  | Article |Article explaining how to utilize the socks5 support|   
+| 1. | Middleware layer | In order to enable socks5 support, several modifications of the `WsTransportClientBuilder` needs to be implemented, described in [issue #1162](https://github.com/paritytech/jsonrpsee/issues/1162) |  
+| 2.  | Socks5 support |enable a jsonrpsee client to proxy connections using a socks5 proxy |   
+
+
+#### Demonstrations:
```
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:74]:
```suggestion
#### Demonstrations
```

File: docs/RFPs/jsonrpsee-proxy-support.md (hunk 10 above)
Comment (takahser) [inline docs/RFPs/jsonrpsee-proxy-support.md:38]:
```suggestion
## Motivation
```

**Review by takahser — CHANGES_REQUESTED** [review]
@flipchan thanks, I'm fine with the contents as well. I just added a couple of stylistic improvement suggestions; feel free to have a look.

**Comment by flipchan** [comment]
Requested changes has been added @takahser

**Review by takahser — APPROVED** [review]

**Review by Noc2 — APPROVED** [review]

**Review by keeganquigley — APPROVED** [review]
Thanks @flipchan this is awesome. Thanks for writing it. And thanks @takahser for edits.

**Pull request merged**